activemodel/lib/active_model/secure_password.rb

```diff
@@ -11,8 +12,24 @@ module SecurePassword
 
     class << self
       attr_accessor :min_cost # :nodoc:
+      attr_writer :cost # :nodoc:
+
+      def cost # :nodoc:
+        min_cost ? BCrypt::Engine::MIN_COST : @cost
+      end
+
+      # Start with cost factor 13 (< ~400ms to compute in Jan 2019) and
+      # double compute cost (increase cost factor by 1) every 18 months.
+      def default_cost # :nodoc:
+        initial_cost = 13
+        days_since_initial_cost = Date.today - Date.parse("2019-01-01")
+        moores_interval_in_days = 1.5 * 365
+
+        initial_cost + days_since_initial_cost.div(moores_interval_in_days)
+      end
```

Comment:
Up for discussion. Idea is to scale with attackers' offline cracking CPU capability, not ours. Hence tracking a Moore's Lawsy thing.

Rarely changes, so calculated and assigned on boot.

Reply:
Hmm.. this feels too aggressive to me. Is this value really still increasing Moorishly?

My superficial research says that in 1999 the recommended value was 8, and at that time CPUs were still outright doubling in performance (so I suspect that the path from 8 to ~13 hasn't even been linear: I think it probably mostly happened in the first few years). If we'd implemented this at 8 in 1999, today it'd be 21, and taking >100 _seconds_ on my laptop. 

(I haven't yet found a direct reference to how long 8 took on a 2000-era CPU, but reading between some lines it sounds like the target was in the 100ms range. For my i7 3.3 GHz CPU, that's still cost=11; 13 is more like 430ms. Either way we're talking an increase of 3-5 over 19 years, way below the ~13 this formula delivers.)

Overall my gut says I care more about local cost than guessing attacker cost: people associate a delay during login with a failure, so I wouldn't want to create enough work for users to perceive a delay. (Especially as the trade-off moves more and more to being explicitly in those terms: what's getting progressively cheaper for an attacker isn't cycles, but parallelism to try many slow hashes at once; meanwhile our countermeasure is to slow down each calculation, but we're thus ultimately left trading visible user-wait-during-login time to protect a stolen password.) If our application is happily running on an ever-shrinking percentage of top-available CPU speed, I understand the desire to have password strength keep pace with the world, but if that means multi-second full-CPU delays during login, that sounds not-so-great.

I really like the idea of automatically rehashing when below-cost passwords are encountered... I wonder if a new "best guess of appropriate cost" function would better live upstream though, setting aside the above questions about exactly what that function looks like.

Reply:
100 seconds yes 😂

Single-core perf has certainly been pretty pokey. Found this analysis (2012): https://preshing.com/20120208/a-look-back-at-single-threaded-cpu-performance/

Re. local cost, we could clamp to `BCrypt::Engine.calibrate(max_local_execution_time)`. Mildly painful to slow down boot to calculate that.

Reply:
Knocked back to a more conservative 5yr-ish doubling period based on 14% yearly increase in single core perf.

Reply:
> Re. local cost, we could clamp to `BCrypt::Engine.calibrate(max_local_execution_time)`. Mildly painful to slow down boot to calculate that.

We're obliged to consider than an attacker might suddenly find themselves a magic CPU that scales differently, but I think we can make stronger assumptions about our own...

`BCrypt::Engine.calibrate(max_local_execution_time / 16) + 4`

Reply:
> but reading between some lines it sounds like the target was in the 100ms range

From what I've understood it was 250ms, but I can't have a reference for that handy either.

Reply:
> I wonder if a new "best guess of appropriate cost" function would better live upstream though

Agreed. IMO the big win here is rehashing which allows Rails and app developers to tune the cost property over time, which was what bcrypt was designed for. 

Future proofing that variable the next 5 years and beyond feels like over-engineering. Besides making a big assumption about CPU performance, making promises beyond what likely will be the maintenance window for this Rails release, I also wonder if it actually makes a difference. Based on [this blog post](https://www.pxdojo.net/2015/08/what-i-learned-from-cracking-4000.html) regarding Ashley Madison's password database (using bcrypt cost 12) it was still guessing for commonly used weak passwords that allowed the author to get 4000 out of 6 million passwords over 5 days time.